### PR TITLE
core: make _sympify reject Basic subclasses

### LINKED ISSUE #20124: Sympify sometimes returns non-Basic
The `_sympify` function is supposed to be for strict sympification:
```julia
In [11]: from sympy.core.sympify import _sympify, sympify                                                                                      

In [12]: sympify(None)                                                                                                                         

In [13]: _sympify(None)                                                                                                                        
---------------------------------------------------------------------------
SympifyError
```
This is useful because `_sympify` is the main way that inputs are validated across the sympy codebase. It shoudl be possible to know after sympifying arguments to a public API that an error will be raised or otherwise a `Basic` instance will be returned. This assumption is implicitly used to guarantee that e.g. `Basic` instances only have `Basic` args.

I have just realised that `_sympify` actually allows `Basic` *subclasses* (rather than instances) to be considered as sympified:
```julia
In [17]: _sympify(Add)                                                                                                                         
Out[17]: sympy.core.add.Add

In [18]: _sympify(sin)                                                                                                                         
Out[18]: sin
```
This is a problem because a `Basic` subclass is not a `Basic` instance and this means that the key invariant that `_sympify` is supposed to guarantee is not necessarily satisfied.

You can see examples of nonsensical objects such as
```julia
In [39]: Integral(x, cos)                                                                                                                      
Out[39]: 
⌠         
⎮ x d(cos)
⌡         
```
Actually doing anything with this object fails:
```julia
In [40]: Integral(x, cos).doit()                                                                                                               
---------------------------------------------------------------------------
TypeError: unsupported operand type(s) for *: 'Symbol' and 'FunctionClass'
```

This happens because in `sympify` the first check if for `__sympy__`:
https://github.com/sympy/sympy/blob/0f260b261b648f2e1624c12e5fbaab971956a405/sympy/core/sympify.py#L336-L338
The `__sympy__` property is added to all `Basic` class by the metaclass `BasicMeta`:
https://github.com/sympy/sympy/blob/0f260b261b648f2e1624c12e5fbaab971956a405/sympy/core/core.py#L67-L71
This means that `Basic` subclasses as well as instances have the attribute but as a property object:
```julia
In [1]: sin(1).__sympy__                                                                                                                       
Out[1]: True

In [2]: sin.__sympy__                                                                                                                          
Out[2]: <property at 0x10c68de00>

In [3]: sin(1).__sympy__ is not None                                                                                                           
Out[3]: True

In [4]: sin.__sympy__ is not None                                                                                                              
Out[4]: True
```

The fix for this is to change this to:
```diff
diff --git a/sympy/core/sympify.py b/sympy/core/sympify.py
index fe91e11628..5b4d775e0a 100644
--- a/sympy/core/sympify.py
+++ b/sympy/core/sympify.py
@@ -334,7 +334,7 @@ def sympify(a, locals=None, convert_xor=True, strict=False, rational=False,
 
     """
     is_sympy = getattr(a, '__sympy__', None)
-    if is_sympy is not None:
+    if is_sympy == True:
         return a
 
     if isinstance(a, CantSympify):
```

That gives:
```julia
In [2]: from sympy.core.sympify import _sympify, sympify                                                                                       

In [3]: sympify(sin)                                                                                                                           
/Users/enojb/current/sympy/sympy/sympy/core/sympify.py:442: SymPyDeprecationWarning: 

String fallback in sympify has been deprecated since SymPy 1.6. Use
sympify(str(obj)) or sympy.core.sympify.converter or obj._sympy_
instead. See https://github.com/sympy/sympy/issues/18066 for more
info.

  SymPyDeprecationWarning(
Out[3]: sin

In [4]: _sympify(sin)                                                                                                                          
---------------------------------------------------------------------------
SympifyError                              Traceback (most recent call last)
<ipython-input-4-fa3ba4c6d40b> in <module>
----> 1 _sympify(sin)

~/current/sympy/sympy/sympy/core/sympify.py in _sympify(a)
    496 
    497     """
--> 498     return sympify(a, strict=True)
    499 
    500 

~/current/sympy/sympy/sympy/core/sympify.py in sympify(a, locals, convert_xor, strict, rational, evaluate)
    417 
    418     if strict:
--> 419         raise SympifyError(a)
    420 
    421     if iterable(a):

SympifyError: SympifyError: sin
```

Making that change leads immediately to brokenness:
```
>>> from sympy.abc import x, y
>>> x + y
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/enojb/current/sympy/sympy/sympy/core/decorators.py", line 251, in _func
    return func(self, other)
  File "/Users/enojb/current/sympy/sympy/sympy/core/decorators.py", line 127, in binary_op_wrapper
    return func(self, other)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 183, in __add__
    return Add(self, other)
  File "/Users/enojb/current/sympy/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/Users/enojb/current/sympy/sympy/sympy/core/operations.py", line 68, in __new__
    c_part, nc_part, order_symbols = cls.flatten(args)
  File "/Users/enojb/current/sympy/sympy/sympy/core/add.py", line 98, in flatten
    from sympy.tensor.tensor import TensExpr
  File "/Users/enojb/current/sympy/sympy/sympy/tensor/tensor.py", line 41, in <module>
    from sympy.combinatorics import Permutation
  File "/Users/enojb/current/sympy/sympy/sympy/combinatorics/__init__.py", line 7, in <module>
    from sympy.combinatorics.polyhedron import (Polyhedron, tetrahedron, cube,
  File "/Users/enojb/current/sympy/sympy/sympy/combinatorics/polyhedron.py", line 830, in <module>
    tetrahedron = Polyhedron(
  File "/Users/enojb/current/sympy/sympy/sympy/combinatorics/polyhedron.py", line 387, in __new__
    obj._faces = FiniteSet(*faces)
  File "/Users/enojb/current/sympy/sympy/sympy/sets/sets.py", line 1768, in __new__
    dargs[i.as_dummy()] = i
  File "/Users/enojb/current/sympy/sympy/sympy/core/basic.py", line 552, in as_dummy
    if not self.has(Symbol):
  File "/Users/enojb/current/sympy/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/Users/enojb/current/sympy/sympy/sympy/core/basic.py", line 1210, in has
    return any(self._has(pattern) for pattern in patterns)
  File "/Users/enojb/current/sympy/sympy/sympy/core/basic.py", line 1210, in <genexpr>
    return any(self._has(pattern) for pattern in patterns)
  File "/Users/enojb/current/sympy/sympy/sympy/core/basic.py", line 1219, in _has
    pattern = _sympify(pattern)
  File "/Users/enojb/current/sympy/sympy/sympy/core/sympify.py", line 498, in _sympify
    return sympify(a, strict=True)
  File "/Users/enojb/current/sympy/sympy/sympy/core/sympify.py", line 419, in sympify
    raise SympifyError(a)
sympy.core.sympify.SympifyError: SympifyError: <class 'sympy.core.symbol.Symbol'>
```
The problem there is that the `Basic` method `has` allows passing in a class or an instance:
```julia
In [1]: exp(x).has(exp)                                                                                                                        
Out[1]: True

In [2]: exp(x).has(x)                                                                                                                          
Out[2]: True
```
I recently committed a change to make `has` use `_sympify` rather than `sympify` (in sympy 1.5 you could do `exp(x).has("x")`). However `has` and probably other methods and function must have been depending on being able to sympify Basic subclasses for years.

Although this would not be a backwards compatible change I think that it should be done anyway. A deprecation warning could be given for the cases where the `__sympy__` attribute is anything other than `True`. That would be enough to flush out any examples in the sympy codebase through the test suite.

An alternative approach that is potentially less disrupting to downstream users would be to change this only for `_sympify` and ensure that `_sympify` is used internally everywhere in sympy. On the other hand using `_sympify` everywhere is potentially more disrupting because it means that string inputs can't be used with the associated functions. I think more users would notice the change to string inputs than the change to sympifying Basic subclasses.

### PR DESCRIPTION
**Pull request by oscarbenjamin**
Previously a Basic subclass could pass through _sympify e.g.:
```
  >>> _sympify(sin) is sin
  True
```
This was problemantic because _sympify is used to validate and prepare
args for many Basic subclasses which should only have Basic args. With
this change using _sympify with a Basic subclass will now raise an
error:
```
  >>> _sympify(sin)
  ...
  SympifyError: SympifyError: sin
```
This should mean that the output of _sympify is guaranteed to be of type
Basic.

#### References to other Issues or PRs

Fixes https://github.com/sympy/sympy/issues/20124

#### Release Notes

NO ENTRY

### CODE REVIEW

**Review by asmeurer — APPROVED** [review]

File: sympy/core/basic.py [hunk 1]
```diff
@@ -1216,7 +1216,8 @@ def _has(self, pattern):
             return any(f.func == pattern or f == pattern
             for f in self.atoms(Function, UndefinedFunction))
 
-        pattern = _sympify(pattern)
+        if not (isinstance(pattern, type) and issubclass(pattern, Basic)):
+            pattern = _sympify(pattern)
         if isinstance(pattern, BasicMeta):
```
Comment (eric-wieser) [inline sympy/core/basic.py:1221]: This first if is essentially mutually exclusive with the second, isn't it? So perhaps it should be changed to an `else`

Comment (oscarbenjamin) [reply sympy/core/basic.py:1221]: Oh, good point. Actually I should have just swapped the order of these.

Comment (oscarbenjamin) [reply sympy/core/basic.py]: I've just moved the call to `_sympify` after the check for `BasicMeta`.

**Comment by oscarbenjamin** [comment]
I was expecting more failures from across the codebase due to this. Maybe there are few errors because the code that is buggy is more likely to be using `sympify` rather than `_sympify`.

If tests pass on this run I might push a temporary commit to test a stronger verification that `_sympify` is always producing `Basic` instances and that `sympify` can only produce instances of `list`, `Basic`, `BasicMeta`, and `NoneType`.

**Comment by oscarbenjamin** [comment]
So it looks like sympify with string input can return anything as it doesn't sympify the result of `parse_expr`:
```julia
In [1]: type(sympify('True'))                                                                                                                  
Out[1]: bool

In [2]: type(sympify(True))                                                                                                                    
Out[2]: sympy.logic.boolalg.BooleanTrue

In [3]: type(sympify('Matrix([[1]])'))                                                                                                         
Out[3]: sympy.matrices.dense.MutableDenseMatrix

In [4]: type(sympify(Matrix([[1]])))                                                                                                           
Out[4]: sympy.matrices.immutable.ImmutableDenseMatrix
```
I'm not sure if that should be considered a bug or not. It seems that `_sympify` only returns `Basic` (in the test suite).

**Pull request merged**